src/test/java/hudson/plugins/git/GitSCMTest.java

```diff
@@ -447,21 +447,95 @@ public void testAvoidRedundantFetchWithHonorRefSpec() throws Exception {
     }
 
+    @Test
+    @Issue("JENKINS-49757")
+    public void testAvoidRedundantFetchWithNullRefspec() throws Exception {
+        String nullRefspec = null;
+        List<UserRemoteConfig> repos = new ArrayList<>();
+        repos.add(new UserRemoteConfig(testRepo.gitDir.getAbsolutePath(), "origin", nullRefspec, null));
+
+        /* Without honor refspec on initial clone */
+        FreeStyleProject projectWithMaster = setupProject(repos, Collections.singletonList(new BranchSpec("master")), null, false, null);
+        if (random.nextBoolean()) {
+            /* Randomly enable shallow clone, should not alter test assertions */
+            CloneOption cloneOptionMaster = new CloneOption(false, null, null);
+            cloneOptionMaster.setDepth(1);
+            ((GitSCM) projectWithMaster.getScm()).getExtensions().add(cloneOptionMaster);
+        }
+
+        // create initial commit
+        final String commitFile1 = "commitFile1";
+        commit(commitFile1, johnDoe, "Commit in master");
+
+        FreeStyleBuild build = build(projectWithMaster, Result.SUCCESS);
+
+        assertRedundantFetchIsSkipped(build, "+refs/heads/*:refs/remotes/origin/*");
+    }
+
+    /*
+     * When initial clone does not honor the refspec and a custom refspec is used
+     * that is not part of the default refspec, then the second fetch is not
+     * redundant and must not be fetched.
+     *
+     * This example uses the format to reference GitHub pull request 553. Other
+     * formats would apply as well, but the case is illustrated well enough by
+     * using the GitHub pull request as an example of this type of problem.
+     */
+    @Test
+    @Issue("JENKINS-49757")
+    public void testRetainRedundantFetch() throws Exception {
+        String refspec = "+refs/heads/*:refs/remotes/origin/* +refs/pull/553/head:refs/remotes/origin/pull/553";
+        List<UserRemoteConfig> repos = new ArrayList<>();
+        repos.add(new UserRemoteConfig(testRepo.gitDir.getAbsolutePath(), "origin", refspec, null));
+
+        /* Without honor refspec on initial clone */
+        FreeStyleProject projectWithMaster = setupProject(repos, Collections.singletonList(new BranchSpec("master")), null, false, null);
+        if (random.nextBoolean()) {
+            /* Randomly enable shallow clone, should not alter test assertions */
+            CloneOption cloneOptionMaster = new CloneOption(false, null, null);
+            cloneOptionMaster.setDepth(1);
+            ((GitSCM) projectWithMaster.getScm()).getExtensions().add(cloneOptionMaster);
+        }
+
+        // create initial commit
+        final String commitFile1 = "commitFile1";
+        commit(commitFile1, johnDoe, "Commit in master");
+
+        /* Create a ref for the fake pull in the source repository */
+        String[] expectedResult = {""};
+        CliGitCommand gitCmd = new CliGitCommand(testRepo.git, "update-ref", "refs/pull/553/head", "HEAD");
```

Comment:
While writing the code, I was thinking how would I proceed with writing a unit test for the same. The solution I had in my mind was to actually clone a remote the git client repository and fetch a pull request reference. 

I didn't imagine to update the reference to point to a pull request. Thanks for showing me!

Reply:
👍 I hadn't discovered the technique until just recently myself.  Plenty to learn for all of us.